machine_learning/vector_ops.hpp

```diff
@@ -0,0 +1,484 @@
+    return out;
+}
+
+/**
+ * Overloaded operator "<<" to print a 1D vector
+ * @tparam T typename of the vector
+ * @param out std::ostream to output
+ * @param A 1D vector to be printed
+ */
+template <typename T>
+std::ostream &operator<<(std::ostream &out, const std::valarray<T> &A) {
+    // Setting output precision to 4 in case of floating point numbers
+    out.precision(4);
+    for(const auto &a : A) { // For every element in the vector.
+        std::cout << a << ' '; // Print element
+    }
+    std::cout << std::endl;
+    return out;
+}
+
+/**
+ * Function to insert element into 1D vector
+ * @tparam T typename of the 1D vector and the element
+ * @param A 1D vector in which element will to be inserted
+ * @param ele element to be inserted
+ * @return new resultant vector
+ */
+template <typename T>
+std::valarray<T> insert_element(const std::valarray <T> &A, const T &ele) {
+    std::valarray <T> B; // New 1D vector to store resultant vector
+    B.resize(A.size() + 1); // Resizing it accordingly
+    for(size_t i = 0; i < A.size(); i++) { // For every element in A
+        B[i] = A[i]; // Copy element in B
+    }
+    B[B.size() - 1] = ele; // Inserting new element in last position
+    return B; // Return resultant vector
+}
+
+/**
+ * Function to remove first element from 1D vector
+ * @tparam T typename of the vector
+ * @param A 1D vector from which first element will be removed
+ * @return new resultant vector
+ */
+template <typename T>
+std::valarray <T> pop_front(const std::valarray<T> &A) {
+    std::valarray <T> B; // New 1D vector to store resultant vector
+    B.resize(A.size() - 1); // Resizing it accordingly
+    for(size_t i = 1; i < A.size(); i ++) { // // For every (except first) element in A 
+        B[i - 1] = A[i]; // Copy element in B with left shifted position
+    }
+    return B; // Return resultant vector
+}
+
+/**
+ * Function to remove last element from 1D vector
+ * @tparam T typename of the vector
+ * @param A 1D vector from which last element will be removed
+ * @return new resultant vector
+ */
+template <typename T>
+std::valarray <T> pop_back(const std::valarray<T> &A) {
+    std::valarray <T> B; // New 1D vector to store resultant vector
+    B.resize(A.size() - 1); // Resizing it accordingly
+    for(size_t i = 0; i < A.size() - 1; i ++) { // For every (except last) element in A 
+        B[i] = A[i]; // Copy element in B
+    }
+    return B; // Return resultant vector
+}
+
+/**
+ * Function to equally shuffle two 3D vectors (used for shuffling training data)
+ * @tparam T typename of the vector
+ * @param A First 3D vector
+ * @param B Second 3D vector
+ */
+template <typename T>
+void equal_shuffle(std::vector < std::vector <std::valarray<T>> >  &A, 
+                   std::vector < std::vector <std::valarray<T>> >  &B) {
+    // If two vectors have different sizes
+    if(A.size() != B.size())
+    {
+        std::cout << "ERROR : Can not equally shuffle two vectors with different sizes: ";
```

Comment:
```suggestion
        std::cerr << "ERROR : Can not equally shuffle two vectors with different sizes: ";
```
Errors are better output to `std::cerr` stream.